cmd/packages.go

```diff
@@ -253,36 +260,72 @@ func packagesExecWorker(userInput string) <-chan error {
 		}
 		defer cleanup()
 
-		catalog, relationships, d, err := syft.CatalogPackages(src, appConfig.Package.Cataloger.ScopeOpt)
-		if err != nil {
-			errs <- fmt.Errorf("failed to catalog input: %w", err)
-			return
+		s := sbom.SBOM{
+			Source: src.Metadata,
 		}
 
-		sbomResult := sbom.SBOM{
-			Artifacts: sbom.Artifacts{
-				PackageCatalog: catalog,
-				Distro:         d,
-			},
-			Relationships: relationships,
-			Source:        src.Metadata,
+		var results []<-chan artifact.Relationship
+		for _, task := range tasks {
+			c := make(chan artifact.Relationship)
+			results = append(results, c)
+
+			go runTask(task, &s.Artifacts, src, c, errs)
+		}
+
+		for relationship := range mergeResults(results...) {
+			s.Relationships = append(s.Relationships, relationship)
 		}
 
 		if appConfig.Anchore.Host != "" {
-			if err := runPackageSbomUpload(src, sbomResult); err != nil {
+			if err := runPackageSbomUpload(src, s); err != nil {
 				errs <- err
 				return
 			}
 		}
 
 		bus.Publish(partybus.Event{
 			Type:  event.PresenterReady,
-			Value: f.Presenter(sbomResult),
+			Value: f.Presenter(s),
 		})
 	}()
 	return errs
 }
 
+func runTask(t task, a *sbom.Artifacts, src *source.Source, c chan<- artifact.Relationship, errs chan<- error) {
+	defer close(c)
+
+	relationships, err := t(a, src)
+	if err != nil {
+		errs <- err
+		return
+	}
+
+	for _, relationship := range relationships {
+		c <- relationship
+	}
+}
+
+func mergeResults(cs ...<-chan artifact.Relationship) <-chan artifact.Relationship {
```

Comment:
my 2 cents: merging channels into a single channel seems more natural then converting it into a slice for the caller. Having the caller iterate through a channel vs a slice seems like a +0 for me. Is the only other simplification about removing the need for a goroutine for `wg.Wait()`?

Reply:
I think that's one minor advantage. My main observation is that the caller doesn't actually have a need for the results to be returned as a channel. So the level of concurrency created here is slightly more than what's motivated by the logical flow.

Reply:
I think we're going to keep this as is on this PR since it was not introduced here and can take a look at editing this flow in the future if it gets more unclear or if we need to keep operating on relationships in a more nuanced way at the command exec level.